### ISSUE TITLE
infinite recursion in Panic_Core

### ISSUE THREAD

**Issue by zsx** [metaeducation/ren-c#691]
Panic_Core [calls IS_ERROR](https://github.com/metaeducation/ren-c/blob/master/src/core/d-crash.c#L190), which in turn calls VAL_TYPE_DEBUG, which might [call panic_at](https://github.com/metaeducation/ren-c/blob/master/src/include/sys-value.h#L214) , and in turn calls back to Panic_Core with the same REBVAL.

I noticed this when testing a new extension.

**Comment by hostilefork** [comment]
Yes, this is rather annoying.  So far when it's happened I've just been putting a breakpoint in Panic_Core so I catch the first one, and avoiding tackling the fundamental issue.

It's easy enough to set a flag and dampen a recursive panic so at least the second call crashes out.  But I was kind of hoping for some more systemic review and improvement.  Today's panic has been a bit of an organic outgrowth of features, throwing together routines that weren't designed with each other in mind.

One thing I was thinking was that the Assert_XXX routines (or whatever) be rethought to return NULL if they have nothing to panic about, otherwise return the thing to panic on.  Then have **panic_if** or something which doesn't panic on NULL.  So more like:

     #define ASSERT_CONTEXT(c) \
         panic_if (Maybe_Context_Defect(c))

What this would mean would be that inside the implementation of panics themselves, they could get a predictive sense about whether trying to work with something would cause a recursive panic...and maybe get extra information.
